File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:

        r = 0
        l = 0
        ans = 0
        char_set = set()
        while(r < len(s)):

            while(s[r] in char_set):
                char_set.remove(s[l])
                l += 1
            
            char_set.add(s[r])

            ans = max(ans , r - l + 1)
            r += 1
        
        return ans

            

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        

        int j = 0 ;

        map<char , int> mp;
        int ans = INT_MIN;
        for(int i = 0 ; i < s.size() ; i++){

            if(mp.find(s[i]) != mp.end()){
                
                while(mp.find(s[i]) != mp.end()){
                    mp.erase(s[j]);
                    j++;
                }
            }
            
            mp[s[i]]++;
            ans = max(ans , i - j + 1);
            
        }

        return ans == INT_MIN ? 0 : ans;
    }
};